packages/hardhat-core/src/internal/hardhat-network/stack-traces/vm-debug-tracer.ts

```diff
@@ -0,0 +1,604 @@
+import { TypedTransaction } from "@ethereumjs/tx";
+import VM from "@ethereumjs/vm";
+import { EVMResult } from "@ethereumjs/vm/dist/evm/evm";
+import { InterpreterStep } from "@ethereumjs/vm/dist/evm/interpreter";
+import Message from "@ethereumjs/vm/dist/evm/message";
+import {
+  EIP2929StateManager,
+  StateManager,
+} from "@ethereumjs/vm/dist/state/interface";
+import { Address, BN, setLengthLeft, toBuffer } from "ethereumjs-util";
+
+import { assertHardhatInvariant } from "../../core/errors";
+import { RpcDebugTracingConfig } from "../../core/jsonrpc/types/input/debugTraceTransaction";
+import { InvalidInputError } from "../../core/providers/errors";
+import { RpcDebugTraceOutput, RpcStructLog } from "../provider/output";
+
+// tslint:disable only-hardhat-error
+
+interface StructLog {
+  depth: number;
+  gas: number;
+  gasCost: number;
+  op: string;
+  pc: number;
+  memory: string[];
+  stack: string[];
+  storage: Record<string, string>;
+  memSize: number;
+  error?: object;
+}
+
+type Storage = Record<string, string>;
+
+interface DebugMessage {
+  structLogs: Array<StructLog | DebugMessage>;
+  to: string;
+  result?: EVMResult;
+}
+
+type NestedStructLogs = Array<StructLog | NestedStructLogs>;
+
+function isStructLog(
+  message: StructLog | DebugMessage | undefined
+): message is StructLog {
+  return message !== undefined && !("structLogs" in message);
+}
+
+const EMPTY_MEMORY_WORD = "0".repeat(64);
+
+export class VMDebugTracer {
+  private _enabled = false;
```

Comment:
I think this enabled/disabled logic isn't necessary for this tracer, as its listeners are injected right before tracing, and removed afterward.

Reply:
To be clear, by this i mean using this boolean

Reply:
Fixed in 7d5a5b90165db3bf266d8891b9adcaebc75ecd36